app/assets/javascripts/actions/story.js

```diff
@@ -86,88 +103,99 @@ export const showHistory = (storyId) =>
       dispatch(sendErrorNotification(error))
-      return dispatch(storyFailure(story.id, error))
+      return dispatch(storyFailure(story.id, error, from))
     }
   }
 
-export const saveStory = (storyId, projectId, options) =>
+export const saveStory = (storyId, projectId, options, from) =>
   async (dispatch, getState, { Story }) => {
     const { stories } = getState();
-    const story = Story.findById(stories, storyId)
 
-    dispatch(setLoadingStory(story.id));
+    const story = Story.findById(Story.withScope(stories, from), storyId);
+
+    dispatch(setLoadingStory(story.id, from));
 
     if (Story.isNew(story)) {
       try {
         const newStory = await Story.post(story._editing, projectId)
 
-        dispatch(addStory(newStory));
+        dispatch(addStory(newStory, from));
 
         return dispatch(sendSuccessNotification(
           I18n.t('messages.operations.success.story.create', { story: story._editing.title })
         ));
       }
       catch (error) {
         dispatch(sendErrorNotification(error))
-        return dispatch(storyFailure(story.id, error))
+        return dispatch(storyFailure(story.id, error, from))
       }
     };
 
     if (story._editing._isDirty) {
       try {
         const updatedStory = await Story.update(story._editing, projectId, options);
 
-        dispatch(updateStorySuccess(updatedStory));
-
+        dispatch(updateStorySuccess(updatedStory, from));
+        
         return dispatch(sendSuccessNotification(
           I18n.t('messages.operations.success.story.save', { story: updatedStory.title })
         ));
       }
       catch (error) {
         dispatch(sendErrorNotification(error))
-        return dispatch(storyFailure(story.id, error))
+        return dispatch(storyFailure(story.id, error, from))
       }
     }
 
-    return dispatch(toggleStory(story.id));
+    return dispatch(toggleStory(story.id, from));
   };
 
-export const deleteStory = (storyId, projectId) =>
+export const deleteStory = (storyId, projectId, from) =>
   async (dispatch, getState, { Story }) => {
-    dispatch(setLoadingStory(storyId))
+    dispatch(setLoadingStory(storyId, from))
     try {
       const { stories } = getState();
-      const storyTitle = Story.findById(stories, storyId).title;
+      const storyTitle = Story.findById(Story.withScope(stories, from), storyId).title;
 
       await Story.deleteStory(storyId, projectId);
 
-      dispatch(deleteStorySuccess(storyId));
+      dispatch(deleteStorySuccess(storyId, from));
 
       return dispatch(sendSuccessNotification(
         I18n.t('messages.operations.success.story.delete', { story: storyTitle })
       ));
     }
     catch (error) {
       dispatch(sendErrorNotification(error))
-      return dispatch(storyFailure(storyId, error))
+      return dispatch(storyFailure(storyId, error, from))
+    }
+  }
+
+export const highlight = storyId =>
+  async (dispatch, getState, {}) => {
+    try {
+      dispatch(updateHighlight(storyId, true));
```

Comment:
Is there the chance any of these fails? They look like deterministic actions, it should not fail.

Also, remember: the `catch` won't catch what is inside `setTimeout` because it's not "awaitable" like that.